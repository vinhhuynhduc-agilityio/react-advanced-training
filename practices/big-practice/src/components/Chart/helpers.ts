import {
  FormattedDataTotalTasksByProjects,
  FormattedDataTotalTasksCompleted,
  ProjectsData,
  TaskData
} from '@/types';

// ag-grid
import {
  AgAreaSeriesOptions,
  AgBarSeriesTooltipRendererParams,
  AgChartOptions,
  AgTooltipRendererResult,
} from 'ag-charts-community';

/**
 * Formats the task data for the chart displaying total tasks completed.
 *
 * @param tasks - Array of TaskData containing information about completed tasks.
 * @returns An object containing:
 *   - formattedData: Array of objects representing months with task counts for each year.
 *   - years: Array of unique years extracted from the task data, sorted in ascending order.
 */
const formatDataForChartTotalTasks = (
  tasks: TaskData[]
): { formattedData: FormattedDataTotalTasksCompleted[]; years: number[] } => {
  // Set to store unique years present in the task data
  const yearsSet = new Set<number>();

  // Object to store task counts for each month
  const months: { [key: string]: FormattedDataTotalTasksCompleted } = {
    Jan: { month: 'Jan' },
    Feb: { month: 'Feb' },
    Mar: { month: 'Mar' },
    Apr: { month: 'Apr' },
    May: { month: 'May' },
    Jun: { month: 'Jun' },
    Jul: { month: 'Jul' },
    Aug: { month: 'Aug' },
    Sep: { month: 'Sep' },
    Oct: { month: 'Oct' },
    Nov: { month: 'Nov' },
    Dec: { month: 'Dec' },
  };

  // Process each task to extract the month and year, and update task counts
  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (completedDate !== 'incomplete') {

      // Extract month and year from the completedDate
      const [, monthStr, yearStr] = completedDate.split(' ');
      const currentYear = new Date().getFullYear();
      const currentCentury = Math.floor(currentYear / 100);
      const year = parseInt(`${currentCentury}${yearStr}`, 10); // Convert yearStr to full year (e.g., "23" -> 2023)
      yearsSet.add(year);

      // Initialize the task count for the month and year if not already present
      if (!months[monthStr][year]) {
        months[monthStr][year] = 0;
      }

      // Increment the task count for the respective month and year
      months[monthStr][year]++;
    }
  });

  // Convert the set of years into a sorted array
  const years = Array.from(yearsSet).sort();

  const formattedData = Object.values(months).map((month) => {
    years.forEach((year) => {
      if (month[year] === undefined) {
        month[year] = 0;
      }
    });
    return month;
  });

  // Return the formatted data and list of years
  return { formattedData, years };
};

const formatDataForChartTotalTasksByProjects = (
  tasks: TaskData[],
  projects: ProjectsData[]
): { formattedData: FormattedDataTotalTasksByProjects[]; years: number[] } => {
  const yearsSet = new Set<number>();

  // Initialize the result with all projects, defaulting years to 0
  const result = projects.reduce((acc, project) => {
    acc[project.projectName] = {
      projectName: project.projectName,
    };
    return acc;
  }, {} as { [key: string]: FormattedDataTotalTasksByProjects });

  // Populate task data
  tasks.forEach((task) => {
    const { projectName, startDate } = task;
    const [, , yearStr] = startDate.split(' ');
    const currentYear = new Date().getFullYear();
    const currentCentury = Math.floor(currentYear / 100);
    const year = parseInt(`${currentCentury}${yearStr}`, 10);

    yearsSet.add(year);

    if (!result[projectName][year]) {
      result[projectName][year] = 0; // Initialize year if not already present
    }

    result[projectName][year]++; // Increment the task count
  });

  // Ensure all projects have values for all years
  const years = Array.from(yearsSet).sort();
  const formattedData = Object.values(result).map((project) => {
    years.forEach((year) => {
      if (!project[year]) {
        project[year] = 0; // Default missing years to 0
      }
    });
    return project;
  });

  return { formattedData, years };
};

const formatDataForChartIndividualEmployee = (
  tasks: TaskData[],
  selectedUserId: string
) => {
  const months: Record<string, { month: string; totalTasksCompleted: number }> = {
    Jan: { month: 'Jan', totalTasksCompleted: 0 },
    Feb: { month: 'Feb', totalTasksCompleted: 0 },
    Mar: { month: 'Mar', totalTasksCompleted: 0 },
    Apr: { month: 'Apr', totalTasksCompleted: 0 },
    May: { month: 'May', totalTasksCompleted: 0 },
    Jun: { month: 'Jun', totalTasksCompleted: 0 },
    Jul: { month: 'Jul', totalTasksCompleted: 0 },
    Aug: { month: 'Aug', totalTasksCompleted: 0 },
    Sep: { month: 'Sep', totalTasksCompleted: 0 },
    Oct: { month: 'Oct', totalTasksCompleted: 0 },
    Nov: { month: 'Nov', totalTasksCompleted: 0 },
    Dec: { month: 'Dec', totalTasksCompleted: 0 },
  };

  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (task.userId === selectedUserId && completedDate !== 'incomplete') {
      const [, monthStr] = completedDate.split(' ');

      months[monthStr as keyof typeof months].totalTasksCompleted++;
    }
  });

  return Object.values(months);
};

const renderTooltipChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const month = params.datum[params.xKey];
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: `<div style='text-align: center; line-height: 1.5'>
              <div>${month}</div>
              <div>${tasksCompleted} tasks completed</div>
            </div>`,
    content: '',
    backgroundColor: '#181d1f',
  };
};

// Default chart configuration
const individualEmployeeProgressOptions: AgChartOptions = {
  title: {
    text: `Individual employee's progress`,
  },
  data: [],
  series: [
    {
      type: 'area',
      xKey: 'month',
      yKey: 'totalTasksCompleted',
      yName: 'Loading...',
      interpolation: { type: 'smooth' },
      tooltip: {
        renderer: renderTooltipChart,
      },
    } as AgAreaSeriesOptions,
  ],
  legend: {
    enabled: true,
    listeners: {
      legendItemClick: (event) => {
        event.preventDefault();
      },
    },
  },
};

const renderTooltipProjectChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: tasksCompleted,
    content: '',
    backgroundColor: '#181d1f',
  };
};

export {
  formatDataForChartTotalTasks,
  formatDataForChartTotalTasksByProjects,
  formatDataForChartIndividualEmployee,
  renderTooltipProjectChart,
  renderTooltipChart,
  individualEmployeeProgressOptions,
};
