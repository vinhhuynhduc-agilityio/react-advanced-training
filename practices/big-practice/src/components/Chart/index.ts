export * from './ChartIndividualEmployeeProgress';
export * from './ChartTotalTasksByProjects';
export * from './ChartTotalTasksCompleted';