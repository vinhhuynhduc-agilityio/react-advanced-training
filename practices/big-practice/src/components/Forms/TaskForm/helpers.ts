// types
import {
  TaskData,
} from '@/types';

const formatDropdownOptions = <T extends { id: string }>(
  data: T[],
  valueKey: keyof T
) =>
  data.map((item) => ({
    id: item.id,
    value: item[valueKey],
  }));

/**
 * Checks if a task name already exists in the task list.
 *
 * @param tasks - Array of existing tasks
 * @param value - New task name to check
 * @returns True if the task name already exists, otherwise false
 */
const isTaskDuplicate = (
  tasks: TaskData[],
  value: string
): boolean => {
  return tasks.some(
    task => task.taskName.toUpperCase() === value.toUpperCase()
  );
};

export {
  formatDropdownOptions,
  isTaskDuplicate
};
