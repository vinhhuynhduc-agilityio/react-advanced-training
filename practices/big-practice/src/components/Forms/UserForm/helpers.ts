import { UserData } from "@/types";

const emailValidation = (defaultEmail: string) => {

  const isEmailDuplicate = (email: string, users: UserData[]) => {
    return users.some(
      (user) => user.email === email && user.email !== defaultEmail
    );
  };

  return { isEmailDuplicate };
};

const readFileAsBase64 = (
  file: File,
  onSuccess: (base64: string) => void
) => {
  const reader = new FileReader();

  reader.readAsDataURL(file);

  reader.onloadend = () => {
    onSuccess(reader.result as string);
  };
};

export {
  emailValidation,
  readFileAsBase64
};
