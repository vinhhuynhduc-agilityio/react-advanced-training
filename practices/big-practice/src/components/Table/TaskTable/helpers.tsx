import { FIELD_TYPE, FieldType } from "@/constant";
import { FieldValue, ProjectsData, TaskData, UserData } from "@/types";

const handleTaskNameChange = (
  value: FieldValue,
  row: TaskData,
  originalTaskNameRef: React.MutableRefObject<string>,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void,
) => {
  const currentValue = originalTaskNameRef.current;
  const newValue = value;

  if (newValue === currentValue) return;

  handleSaveSelect(FIELD_TYPE.TASK_NAME as FieldType, newValue, row);
};

const handleProjectChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const currentValue = row.projectId;
  const newValue = (value as ProjectsData).id;

  if (newValue === currentValue) return;

  handleSaveSelect(FIELD_TYPE.PROJECT as FieldType, value, row);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          projectId: newValue,
          projectName: (value as ProjectsData).projectName,
        }
        : task
    )
  );
};

const handleUserChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  updateEarningsForUsers: (oldUserId: string, newUserId: string, currency: number, status: boolean) => void,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const { userId, currency, status } = row;
  const oldUserId = userId;
  const newUserId = (value as UserData).id;

  if (newUserId === oldUserId) return;

  handleSaveSelect(FIELD_TYPE.USER as FieldType, value, row);
  updateEarningsForUsers(oldUserId, newUserId, currency, status);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          userId: newUserId,
          fullName: (value as UserData).fullName,
        }
        : task
    )
  );
};

const handleStatusChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  updateEarningsOnStatusChange: (userId: string, currency: number, status: boolean) => void,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const status = value as boolean;
  const { currency, userId } = row;
  const completedDate = status
    ? new Date().toLocaleDateString('en-GB', {
      day: '2-digit',
      month: 'short',
      year: '2-digit',
    } as Intl.DateTimeFormatOptions)
    : 'incomplete';

  handleSaveSelect(FIELD_TYPE.STATUS as FieldType, { status, completedDate }, row);
  updateEarningsOnStatusChange(userId, currency, status);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          status,
          completedDate,
        }
        : task
    )
  );
};

export {
  handleTaskNameChange,
  handleProjectChange,
  handleUserChange,
  handleStatusChange
};
