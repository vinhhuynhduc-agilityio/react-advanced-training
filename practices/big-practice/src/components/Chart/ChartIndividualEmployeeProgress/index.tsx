import {
  memo,
  useMemo,
} from 'react';

// ag-grid
import { AgCharts } from 'ag-charts-react';
import { AgChartOptions } from 'ag-charts-community';

// helpers
import { formatDataForChartIndividualEmployee, individualEmployeeProgressOptions } from '@/components/Chart/helpers';

// component
import { Spinner } from '@/components/common';

// types
import { TaskData, UserData } from '@/types';

interface ChartIndividualEmployeeProgressProps {
  selectedUserId: string | null;
  isLoading: boolean;
  users: UserData[];
  tasks: TaskData[];
};

export const ChartIndividualEmployeeProgress: React.FC<ChartIndividualEmployeeProgressProps> = memo(
  ({
    selectedUserId,
    isLoading,
    users,
    tasks
  }) => {

    const selectedUser = useMemo(() => {
      if (users.length === 0) return null;
      if (!selectedUserId) return users[0];
      return users.find((user) => user.id === selectedUserId) || null;
    }, [users, selectedUserId]);

    const memoizedOptions = useMemo(() => {
      const formattedData = selectedUser
        ? formatDataForChartIndividualEmployee(tasks, selectedUser.id)
        : [];

      return {
        ...individualEmployeeProgressOptions,
        data: formattedData,
        series: [
          {
            ...(individualEmployeeProgressOptions.series?.[0] || {}),
            yName: selectedUser?.fullName,
          },
        ],
      } as AgChartOptions;
    }, [tasks, selectedUser]);

    if (isLoading) {
      return (
        <div
          className="flex-1 mr-4 bg-white border border-customBorder"
          role="figure"
          aria-label="Loading chart: Individual employee progress"
        >
          <Spinner />
        </div>
      );
    }

    return (
      <div
        id={`chart-employee-progress-${selectedUser?.id || "unknown"}`}
        role="figure"
        aria-label={`Chart showing progress of ${selectedUser?.fullName || "unknown employee"}`}
        className="flex-1 mr-4 bg-white border border-customBorder"
      >
        <AgCharts options={memoizedOptions} />
      </div>
    );
  }
);
