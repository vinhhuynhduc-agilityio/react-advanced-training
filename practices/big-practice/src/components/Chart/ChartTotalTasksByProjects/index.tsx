import {
  useMemo,
  memo
} from 'react';
import { AgCharts } from 'ag-charts-react';

// helpers
import { formatDataForChartTotalTasksByProjects, totalTasksByProjectsOption } from '@/components/Chart/helpers';

// component
import { Spinner } from '@/components/common';

// types
import { ProjectsData, TaskData } from '@/types';


interface ChartTotalTasksByProjectsProps {
  isLoading: boolean;
  tasks: TaskData[];
  projects: ProjectsData[]
}

export const ChartTotalTasksByProjects: React.FC<ChartTotalTasksByProjectsProps> = memo(
  ({ isLoading, tasks, projects }) => {

    const formattedData = useMemo(
      () => formatDataForChartTotalTasksByProjects(tasks, projects),
      [tasks, projects]
    );

    const memoizedOptions = useMemo(() => ({
      ...totalTasksByProjectsOption,
      data: formattedData,
    }), [formattedData]);

    if (isLoading) {
      return (
        <div
          className="flex-1 bg-white border border-customBorder"
          role="figure"
          aria-label="Loading chart: Total tasks by projects"
        >
          <Spinner />
        </div>
      );
    }

    return (
      <div
        id="chart-tasks-by-projects"
        role="figure"
        aria-label="Chart showing total tasks by projects"
        className="flex-1 bg-white border border-customBorder"
      >
        <AgCharts options={memoizedOptions} />
      </div>
    );
  }
);
