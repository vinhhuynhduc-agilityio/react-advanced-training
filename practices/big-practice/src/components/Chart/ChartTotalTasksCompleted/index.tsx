import {
  useMemo,
  memo
} from 'react';
import { AgCharts } from 'ag-charts-react';

// component
import { Spinner } from '@/components/common';

// helpers
import { formatDataForChartTotalTasks, totalTasksCompletedOptions } from '@/components/Chart/helpers';

// types
import { TaskData } from '@/types';

interface ChartTotalTasksCompletedProps {
  isLoading: boolean;
  tasks: TaskData[]
};

export const ChartTotalTasksCompleted: React.FC<ChartTotalTasksCompletedProps> = memo(
  ({ isLoading, tasks }) => {

    const formattedData = useMemo(() => formatDataForChartTotalTasks(tasks), [tasks]);

    const memoizedOptions = useMemo(() => ({
      ...totalTasksCompletedOptions,
      data: formattedData,
    }), [formattedData]);

    if (isLoading) {
      return (
        <div
          className="flex-1 bg-white border border-customBorder h-[302px]"
          role="figure"
          aria-label="Loading chart: Total tasks completed"
        >
          <Spinner />
        </div>
      );
    }

    return (
      <div
        id="chart-total-tasks-completed"
        role="figure"
        aria-label="Chart showing total tasks completed"
        className="flex-1 bg-white border border-customBorder h-[302px]"
      >
        <AgCharts options={memoizedOptions} />
      </div>
    );
  }
);
