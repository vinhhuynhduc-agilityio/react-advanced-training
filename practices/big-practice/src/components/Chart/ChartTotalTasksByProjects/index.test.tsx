import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

// helpers
import { formatDataForChartTotalTasksByProjects } from '@/components/Chart/helpers';
import { ChartTotalTasksByProjects } from '.';

// Mock AgCharts component to simulate its rendering
jest.mock('ag-charts-react', () => ({
  AgCharts: () => <div data-testid="total-tasks-chart"></div>,
}));

// Mock helper functions
jest.mock('@/components/Chart/helpers', () => ({
  ...jest.requireActual('@/components/Chart/helpers'),
  formatDataForChartTotalTasksByProjects: jest.fn(),
}));

// Mock data
import { mockTasks, mockProjects } from '@/mocks';

describe('ChartTotalTasksByProjects Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the chart component', () => {
    const { container } = render(
      <ChartTotalTasksByProjects isLoading={false} tasks={mockTasks} projects={mockProjects} />
    );

    expect(screen.getByTestId('total-tasks-chart')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('calls formatDataForChartTotalTasksByProjects with correct data', () => {
    render(
      <ChartTotalTasksByProjects isLoading={false} tasks={mockTasks} projects={mockProjects} />
    );

    expect(formatDataForChartTotalTasksByProjects).toHaveBeenCalledWith(mockTasks, mockProjects);
  });

  it('does not crash if there are no tasks or projects', () => {
    const { container } = render(
      <ChartTotalTasksByProjects isLoading={false} tasks={[]} projects={[]} />
    );

    expect(screen.getByTestId('total-tasks-chart')).toBeInTheDocument();
    expect(formatDataForChartTotalTasksByProjects).toHaveBeenCalledWith([], []);
    expect(container).toMatchSnapshot();
  });

  it('renders a spinner when isLoading is true', () => {
    const { container } = render(
      <ChartTotalTasksByProjects isLoading={true} tasks={mockTasks} projects={mockProjects} />
    );

    const spinner = container.querySelector(
      '.w-12.h-12.border-4.border-gray-300.border-t-blue-500.rounded-full.animate-spin'
    );
    expect(spinner).toBeInTheDocument();

    const mockedChart = screen.queryByTestId('total-tasks-chart');
    expect(mockedChart).not.toBeInTheDocument();
  });
});
