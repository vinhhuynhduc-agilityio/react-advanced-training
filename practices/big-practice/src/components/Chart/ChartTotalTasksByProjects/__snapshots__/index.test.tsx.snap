// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTotalTasksByProjects Component does not crash if there are no tasks or projects 1`] = `
<div>
  <div
    aria-label="Chart showing total tasks by projects"
    class="flex-1 bg-white border border-customBorder"
    id="chart-tasks-by-projects"
    role="figure"
  >
    <div
      data-testid="total-tasks-chart"
    />
  </div>
</div>
`;

exports[`ChartTotalTasksByProjects Component renders the chart component 1`] = `
<div>
  <div
    aria-label="Chart showing total tasks by projects"
    class="flex-1 bg-white border border-customBorder"
    id="chart-tasks-by-projects"
    role="figure"
  >
    <div
      data-testid="total-tasks-chart"
    />
  </div>
</div>
`;
